### ISSUE TITLE
Add an autocommit property to sqlite3.Connection with a PEP 249 compliant manual commit mode and migrate

### ISSUE THREAD

**Issue by 53cd72ed-7768-4f34-9566-9995ab0e9dae** [python/cpython#83638]
BPO | [39457](https://bugs.python.org/issue39457)
--- | :---
Nosy | @malemburg, @bitdancer, @maggyero, @erlend-aasland

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2020-01-26.19:17:09.219>
labels = ['type-feature', 'library', '3.9']
title = 'Add an autocommit property to sqlite3.Connection with a PEP 249 compliant manual commit mode and migrate'
updated_at = <Date 2021-01-06.13:21:56.616>
user = 'https://github.com/maggyero'
```

bugs.python.org fields:
```python
activity = <Date 2021-01-06.13:21:56.616>
actor = 'zzzeek'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2020-01-26.19:17:09.219>
creator = 'maggyero'
dependencies = []
files = []
hgrepos = []
issue_num = 39457
keywords = []
message_count = 12.0
messages = ['360732', '360734', '360735', '384383', '384397', '384414', '384417', '384421', '384425', '384485', '384499', '384501']
nosy_count = 7.0
nosy_names = ['lemburg', 'ghaering', 'r.david.murray', 'zzzeek', 'maggyero', 'erlendaasland', 'james.oldfield']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue39457'
versions = ['Python 3.9']
```

</p></details>

* PR: gh-93823

**Comment by erlend-aasland** [comment]
FTR, I've changed the formatting of https://github.com/python/cpython/issues/83638#issuecomment-1093853902 so it does not eat up multiple screenfuls using heading formatting for every line.

**Comment by erlend-aasland** [comment]
@zzzeek, would the solution outlined by MAL in https://github.com/python/cpython/issues/83638#issuecomment-1093853924 solve SQLAlchemy's problems? If there is any confusion regarding an attribute named `autocommit`, it can be resolved in the docs.

We could add the new attribute and introduce deprecation warnings for `isolation_level` in 3.12. I don't think we should remove `isolation_level` anytime soon; a possibility is to keep it forever as deprecated.

This issue should be focused only on adding the new property, and not on the deprecation of isolation_level. Resolving it should be possible using a single PR.

**Comment by zzzeek** [comment]
@erlend-aasland that comment looks like it wants to solve multiple things at once, which is, adding an autocommit attribute, but also changing the BEGIN behavior that I noted in #54133 when it states "start a new transaction when the connection is opened" (Which I assume applies to the case that "autocommit" would be passed to the sqlite3 connect() method).

I haven't read the whole set of comments above but if you added an autocommit parameter that's fine, but it should be readable also, which means if I have a regular `sqlite3.connect()` connection, autocommit should be False, not None.      I think syncing autocommit and isolation_level parameters together, for as long as isolation_level is present, would be least confusing, that is, if I set isolation_level to None, that implies `connection.autocommit` now returns True.

I guess you also need to make another new parameter to allow control of "DEFERRED", "IMMEDIATE", "EXCLUSIVE".   SQLAlchemy doesn't have direct API for these parameters.    our API right now does what you see below, so it would not be difficult to change it for an .autocommit parameter.    A bigger issue would be any default behavioral changes, like, files get locked sooner, things like that, we'd get a lot of user complaints blaming us for that sort of thing if it just changed.

```py
# sqlalchemy's sqlite3 isolation level pseudocode
_isolation_lookup = {"READ UNCOMMITTED": 1, "SERIALIZABLE": 0}

def set_isolation_level(self, dbapi_connection, level):
    """set the isolation level for a sqlite3 connection"""

    if level == "AUTOCOMMIT":
        dbapi_connection.isolation_level = None
    else:
        # i believe this is eqvuialent to DEFERRED
        dbapi_connection.isolation_level = ""

        # convert from "READ UNCOMMITTED" or "SERIALIZABLE" to an int
        # read_uncommitted level
        int_level = _isolation_lookup[level]
        cursor = dbapi_connection.cursor()
        cursor.execute("PRAGMA read_uncommitted = %d" % int_level)
        cursor.close()


```

**Comment by erlend-aasland** [comment]
> that comment looks like it wants to solve multiple things at once, which is, adding an autocommit attribute, but also changing the BEGIN behavior [...]

Yes, that is also how I interpret MAL's comment. As I understand it, the only way to make the sqlite3 extension module behave as expected (PEP 249 compliant), is to control behaviour via a new attribute. So, yes, it would solve multiple things at once.

The alternative is to change the sqlite3 extension module's behaviour wrt. `isolation_level`, which would be backwards incompatible in any kind of way.

> [...] if you added an autocommit parameter that's fine, but it should be readable also [...]

+1

> [...] which means if I have a regular `sqlite3.connect()` connection, autocommit should be False, not None. I think syncing autocommit and isolation_level parameters together, for as long as isolation_level is present, would be least confusing, that is, if I set isolation_level to None, that implies `connection.autocommit` now returns True.

How could we introduce new behaviour if they are to be sync'ed? `autocommit` would just be another way to control `isolation_level`, so we end up keeping the current behaviour, which is buggy.

I think MAL's suggestion to how a hypothetical `autocommit` attribute could work, sounds good to me. Setting it to `None` to fall back to pre 3.12 behaviour also sounds good to me.

I'm not sure how the interplay between `autocommit` and `isolation_level` should be, yet.

> I guess you also need to make another new parameter to allow control of "DEFERRED", "IMMEDIATE", "EXCLUSIVE". 

Yeah, we could add such an attribute. For example `.transaction_control`?

> A bigger issue would be any default behavioral changes, like, files get locked sooner, things like that, we'd get a lot of user complaints blaming us for that sort of thing if it just changed.

Quoting the [SQLite locking docs](https://www.sqlite.org/lockingv3.html):

> Note that the BEGIN command does not acquire any locks on the database. After a BEGIN command, a SHARED lock will be acquired when the first SELECT statement is executed. A RESERVED lock will be acquired when the first INSERT, UPDATE, or DELETE statement is executed. No EXCLUSIVE lock is acquired until either the memory cache fills up and must be spilled to disk or until the transaction commits. In this way, the system delays blocking read access to the file file until the last possible moment.

Also:

> The SQL command "COMMIT" does not actually commit the changes to disk. It just turns autocommit back on.

But yeah, we need to keep such behavioural changes in mind.

**Comment by zzzeek** [comment]
> > that comment looks like it wants to solve multiple things at once, which is, adding an autocommit attribute, but also changing the BEGIN behavior [...]
> 
> Yes, that is also how I interpret MAL's comment. As I understand it, the only way to make the sqlite3 extension module behave as expected (PEP 249 compliant), is to control behaviour via a new attribute. So, yes, it would solve multiple things at once.
> 
> The alternative is to change the sqlite3 extension module's behaviour wrt. `isolation_level`, which would be backwards incompatible in any kind of way.
> 
> > [...] if you added an autocommit parameter that's fine, but it should be readable also [...]
> 
> +1
> 
> > [...] which means if I have a regular `sqlite3.connect()` connection, autocommit should be False, not None. I think syncing autocommit and isolation_level parameters together, for as long as isolation_level is present, would be least confusing, that is, if I set isolation_level to None, that implies `connection.autocommit` now returns True.
> 
> How could we introduce new behaviour if they are to be sync'ed? `autocommit` would just be another way to control `isolation_level`, so we end up keeping the current behaviour, which is buggy.
> 
> I think MAL's suggestion to how a hypothetical `autocommit` attribute could work, sounds good to me. Setting it to `None` to fall back to pre 3.12 behaviour also sounds good to me.

I wouldn't link the "autocommit" flag to be hardcoded to the "new" behavior, basically, if we are talking about when sqlite3 emits BEGIN, whether it's at the first SQL statement or it's at the first DML (or non DDL, im not sure how that works), that should be a *separate* argument which selects among different "autobegin" styles.   that way you can add "autocommit" and have it work consistently for everyone right up front without it ever changing, and you have the "sqlite3 autobegin behavior" thing separate, where you can eventually change its default.

> 
> I'm not sure how the interplay between `autocommit` and `isolation_level` should be, yet.
> 
> > I guess you also need to make another new parameter to allow control of "DEFERRED", "IMMEDIATE", "EXCLUSIVE".
> 
> Yeah, we could add such an attribute. For example `.transaction_control`?

More than that is needed.  I think you should make a matrix of *all* transactional behaviors and work out combinations of parameters for all of them, combining the concepts of "autocommit", "autobegin (when BEGIN is emitted)", "transaction control" (what keyword is included after BEGIN), and I would include "read_committed" as well.  then work out an API that cleanly handles all combinations.  I dont think SQLite's "BEGIN only on DML/non DDL" feature should be removed because certainly people rely upon it and if you change the default someday, that's fine but people will likely ask for the old behavior.



> 
> > A bigger issue would be any default behavioral changes, like, files get locked sooner, things like that, we'd get a lot of user complaints blaming us for that sort of thing if it just changed.
> 
> Quoting the [SQLite locking docs](https://www.sqlite.org/lockingv3.html):
> 
> > Note that the BEGIN command does not acquire any locks on the database. After a BEGIN command, a SHARED lock will be acquired when the first SELECT statement is executed. A RESERVED lock will be acquired when the first INSERT, UPDATE, or DELETE statement is executed. No EXCLUSIVE lock is acquired until either the memory cache fills up and must be spilled to disk or until the transaction commits. In this way, the system delays blocking read access to the file file until the last possible moment.

beyond locking issues (which I thought was the original rationale for this behavior?)  note also that DDL with sqlite3 is not transactional right now.  so if you change that behavior, all the scripts that connect() and then execute() DDL statements without calling commit(), assuming the DDL is now persistent, will break.

**Comment by erlend-aasland** [comment]
> I wouldn't link the "autocommit" flag to be hardcoded to the "new" behavior [...]

Well, this issue is about adding `autocommit` with PEP 249 compliant behaviour. This means you are -1 for this issue? To try and keep these discussions focused, I suggest opening another issue to discuss the alternative solution you are describing.

**Comment by zzzeek** [comment]
> > I wouldn't link the "autocommit" flag to be hardcoded to the "new" behavior [...]
> 
> Well, this issue is about adding `autocommit` with PEP 249 compliant behaviour. This means you are -1 for this issue? 

I am fully +1 for adding a pep-249 compliant .autocommit attribute.   I am only indicating that the aspect of the sqlite3 driver regarding it's already non-pep-249 compliant treatment of the BEGIN statement, what we see in 54133 , is a separate issue. 

Here is one proposal that meets both the 54133 and this issue:

* autocommit = True  : sqlite3 driver does not emit BEGIN.   commit() and rollback() pass silently
* autocommit = False : this is the default.   sqlite3 driver emits BEGIN at the start of a "transaction".  however "start of a transaction" here has two modes, legacy and pep-249 compliant, controlled as: 
     * sqlite3_legacy_begin_behavior = True:  this is the default in the near term.  when autocommit = False, emits BEGIN only after DML (or DQL), but not DDL.  DDL remains autocommit in legacy mode
     * sqlite3_legacy_begin_behavior = False:  will become the default later.   when autocommit = False, emits BEGIN immediately as soon as any SQL is emitted on the connection

If you think I need to open a new, competing issue that asks that the .autocommit public interface, and the internal legacy behavior  be decoupled, I'll do that, but it seems unnecessary as it would fragment the discussion.

**Comment by erlend-aasland** [comment]
> I am fully +1 for adding a [pep-249](https://peps.python.org/249) compliant .autocommit attribute. I am only indicating that the aspect of the sqlite3 driver regarding it's already non-[pep-249](https://peps.python.org/249) compliant treatment of the BEGIN statement, what we see in https://github.com/python/cpython/issues/54133 , is a separate issue.

They are separate issues in theory, but AFAICS inseparable in practise: because of backwards compatibility, we _cannot_ resolve #54133 without introducing a completely new way to control transaction behaviour, and the proposed new way of controlling transaction behaviour, is this issue.

It seems we agree on the main points:

- Keep a way of selecting the old behaviour
- Introduce a way of selecting PEP 249 compliant behaviour via an `autocommit` attribute

Good!

For the `autocommit` API, we also seem to agree on one value:

- `autocommit=True`: everyone agrees it should behave as described by MAL?
- `autocommit=False`: no agreement

My thoughts are:

I would prefer to keep the new sqlite3 API to a minimum, so I would like to avoid multiple attributes. Right now, I'm inclined to something like this:

`autocommit=False` implies strict PEP 249 compliance, using BEGIN DEFERRED to start transactions. If the user wants anything else, use either the old API (`isolation_level`), or `autocommit=True` and control transactions explicitly[^1].

I don't think it is a good idea to create a matrix of all possible transaction handling (savepoint, etc.) and provide APIs for those; I'm afraid it will hit us in the future. I think a better approach is to provide a new and minimal API in 3.12. If that is not enough, we can expand the API in the future.

For selecting the old behaviour, maybe introducing a constant is better than using `None`? `autocommit = sqlite3.DEPRECATED_TRANSACTION_HANDLING` (bikeshedding welcome).


[^1]: for example `execute("BEGIN IMMEDIATE")`, etc.

**Comment by zzzeek** [comment]
my recent experience with constants and pep-484 typing is that they work better when they are off of a Python ``Enum`` object because then you can type the value of `autocommit` as `Literal[True, False, SQlite3Constants.DEPRECATED_TRANSACTION_HANDLING]`.    it's probably better than None.

also I agree savepoint doesn't belong in a feature matrix here, I was only referring to the presence of DEFERRED /IMMEDIATE / EXCLUSIVE.     I dont use those variants myself but since they are part of SQLite's own BEGIN syntax it would still be more standard for there to be some attribute somewhere which allows this to be controllable within the scope of sqlite3's "autobegin" functionality.

**Comment by erlend-aasland** [comment]
> my recent experience with constants and [pep-484](https://peps.python.org/484) typing is that they work better when they are off of a Python `Enum` object because then you can type the value of `autocommit` as `Literal[True, False, SQlite3Constants.DEPRECATED_TRANSACTION_HANDLING]`. it's probably better than None.

Good point. So we agree on some kind of constant.

> also I agree savepoint doesn't belong in a feature matrix here, I was only referring to the presence of DEFERRED /IMMEDIATE / EXCLUSIVE. I dont use those variants myself but since they are part of SQLite's own BEGIN syntax it would still be more standard for there to be some attribute somewhere which allows this to be controllable within the scope of sqlite3's "autobegin" functionality.

Ok, so my earlier suggestion about providing a single new attribute to control the BEGIN statement should be sufficient? Another suggestion might be to reuse `isolation_mode` for this, but I'm afraid that will only cause more confusion.

**Comment by erlend-aasland** [comment]
I created a proof-of-concept branch for this over at my fork:

- https://github.com/erlend-aasland/cpython/tree/sqlite-autocommit
- https://github.com/erlend-aasland/cpython/pull/35

As of now, it does not work with the context manager, and probably some other stuff. I hasn't got a constant for compat mode either. Feel free to play with it.

**Comment by erlend-aasland** [comment]
> agree the ship has sailed on `isolation_mode` and also these constants really dont resemble at all the traditional notion of "transaction isolation"; SQLite's own documentation does not use this term for them (it doesn't seem to have a term, which is unfortunate).

What about `transaction_behaviour`? It aligns pretty well with the [SQLite docs](https://sqlite.org/lang_transaction.html):

> Transactions can be DEFERRED, IMMEDIATE, or EXCLUSIVE. The default transaction behavior is DEFERRED.

If it helps typing, we could add constants for each of those.

```python
cx.transaction_behaviour = sqlite3.DEFERRED
```

**Comment by erlend-aasland** [comment]
I suggest we create a topic on Discourse to try to raise more awareness on this change.

Is `autocommit=False` equals strict PEP 249 compliancy is a hard showstopper for you, @zzzeek?

**Comment by zzzeek** [comment]
> I suggest we create a topic on Discourse to try to raise more awareness on this ~change~ feature. **EDIT**: [Posted here](https://discuss.python.org/t/introducing-a-pep-249-compliant-transaction-behaviour-to-sqlite3/16516?u=erlendaasland).
> 
> Is `autocommit=False` equals strict PEP 249 compliancy is a hard showstopper for you, @zzzeek?

nah.   I can work with whatever, I was just suggesting.     A sudden change in default behavior, and even with the old behavior no longer available, might be inconvenient on this end, but even then, we are pretty much locking out any patterns where "there's no transaction" unless DBAPI level autocommit is requested, so we could adapt and maybe even without difficulty.

**Comment by erlend-aasland** [comment]
I added docs to my WIP branch, and created a draft PR against the CPython repo. Please try it out, and chime in with comments regarding stuff that I forgot :)

- #93823

**Comment by erlend-aasland** [comment]
Although I've marked `isolation_level` as deprecated in the docs of gh-93823, I think we should create a new PR for adding the deprecation warnings to `isolation_level`. Possibly also as a separate issue.

**Comment by erlend-aasland** [comment]
> * the context manager: do as now, and just commit (or rollback) if there is an open transaction?

For now, the `autocommit` attribute does not affect how the context manager behaves. Suggesting to add an implicit BEGIN after the implicit COMMIT or ROLLBACK, if `autocommit=False`.

> * `autocommit=True`, followed by `execute("BEGIN")`, followed by `autocommit=False`: commit and begin? keep the transaction open? Leaning towards the former.

I've implemented the former in the PR.

**Comment by erlend-aasland** [comment]
Following up this concern:

> I guess you also need to make another new parameter to allow control of "DEFERRED", "IMMEDIATE", "EXCLUSIVE". SQLAlchemy doesn't have direct API for these parameters.

I suggest we keep gh-93823 focused on the autocommit attribute only.

It is still very early in the 3.12 dev phase; we have lots of time to figure out how, and if, we should extend the new API.

**Comment by erlend-aasland** [comment]
> Although I've marked `isolation_level` as deprecated in the docs of [gh-93823](https://github.com/python/cpython/pull/93823), I think we should create a new PR for adding the deprecation warnings to `isolation_level`. Possibly also as a separate issue.

FTR, I've removed the docs deprecation of `isolation_level` from gh-93823; a slower path towards a new default is probably better for everyone involved. For 3.12, this means:

- gh-93823 introduces `autocommit` as a parameter to `connect()` and as an attribute to `Connection`
- the current (3.11) behaviour is still the default behaviour
- `isolation_level` is _not_ deprecated

Then, emit deprecations in 3.14, and finally in 3.16 change the default behaviour to PEP 249-compliant transaction control.

**Comment by erlend-aasland** [comment]
I've merged gh-83638, hopefully in time for the next 3.12 alpha release. IMO, we can handle the possible deprecation of `isolation_level` and how to change the `autocommit` default in a follow-up issue.

Please give it a try!

Thanks to everyone involved here and on previous issues (and on db-sig), for helping to carve out this functionality.

**Comment by erlend-aasland** [comment]
FTR, PEP-249 was updated by python/peps#2887

**Comment by coleifer** [comment]
@erlend-aasland - can you summarize how this new property plays with the previously-supported method of setting `conn.isolation_level = None` to disable the python driver's transaction management? In peewee we run things in autocommit mode across sqlite3/mysql/postgres, and currently use `isolation_level=None` to perform this for Sqlite.

Am I correct that it should be sufficient to check `sys.version_info` for Python 3.12, and in those cases to set `autocommit=True` rather than use `isolation_level`?

**Comment by erlend-aasland** [comment]
@coleifer, can you take a look at the [3.12 docs](https://docs.python.org/3.12/library/sqlite3.html#transaction-control), where this is (or at least should be) thoroughly explained; if you feel more explanation is needed, could you please let me know?

**Comment by zzzeek** [comment]
> 
> Then, emit deprecations in 3.14, and finally in 3.16 change the default behaviour to PEP 249-compliant transaction control.

is 3.16 still the target release for the switch, and will legacy transactions mode be removed in 3.16 at the same time or will it still remain available?
